compiler/qsc_partial_eval/src/management.rs

```diff
@@ -9,20 +11,23 @@ use qsc_eval::{
     val::{Qubit, Result, Value},
 };
 use qsc_rir::rir::{BlockId, CallableId, VariableId};
+use rustc_hash::FxHashSet;
 
 /// Manages IDs for resources needed while performing partial evaluation.
 #[derive(Default)]
 pub struct ResourceManager {
     qubits_in_use: Vec<bool>,
     qubit_id_map: IndexMap<usize, usize>,
+    qubit_tracker: FxHashSet<Rc<Qubit>>,
     next_callable: CallableId,
     next_block: BlockId,
     next_result_register: usize,
     next_var: usize,
 }
 
 impl ResourceManager {
-    pub fn map_qubit(&self, q: Qubit) -> usize {
+    pub fn map_qubit(&self, q: &Weak<Qubit>) -> usize {
+        let q = q.upgrade().expect("qubit should still be alive");
```

Comment:
what's the difference between "qubit should still be alive", "qubit should be checked for liveness before call into backend", "qubit used after release", and the various other `expect()` messages? Are they trying to convey nuanced, different things, or is it just different ways of saying the same thing?

I ask because this is a very subtle/nuanced change -- these messages refer to some implicit functionality in this part of our code and make an analogy to an advanced Rust concept. 

A docstring that describes how we are using `Weak` and how that correlates to our qubit liveness model would be very helpful in the future I think

Reply:
I hope the `QubitRef` struct introduced helps with this. Now there is one spot that panics, and the callers of that function can be found with easy search (or from stack traces in case this happens in practice).